modules/local/locidex_select.nf

```diff
@@ -0,0 +1,276 @@
+        }
+        // Create output summary data for a locidex allele database if an allele scheme is passed in directly
+        def jsonSlurper = new JsonSlurper()
+        String json_data = config.text
+        def locidex_config_data = jsonSlurper.parseText(json_data)
+        validate_locidex_db(locidex_config_data, params.allele_scheme)
+        write_config_data(locidex_config_data, output_config)
+        scheme = params.allele_scheme // reset the schem path to the passed allele scheme
+        paired_p = true
+        output_config
+        paired_p
+        scheme
+
+    }else{
+
+        // Tokenize the "top_hit" or species value to identify all relevant match parts of the string
+        def species_data = top_hit.split('_|\s').collect{ it.toLowerCase() }
+
+        // De-serialize the manifest file from the database location
+        def jsonSlurper = new JsonSlurper()
+        String json_data = manifest.text
+        def allele_db_data = jsonSlurper.parseText(json_data)
+        def allele_db_keys = allele_db_data.keySet() as String[]
+
+        // Tokenize all database keys for lookup of species top hit in the database names
+        def shortest_entry = Integer.MAX_VALUE
+        def databases = allele_db_keys.collect{
+                            key ->  // Tokenize the database key and normalize the values into lowercase
+                                    def db_tokens = key.split('_|\s').collect{ it.toLowerCase() }
+
+                                    // Get the shortest string from the tokenized string (db_tokens), db_tokens.min
+                                    // iterates over the tokens, and returns the shortest entry based on its size. Size
+                                    // is called again to to get the length of the shortest string, then Math.min is called
+                                    // to compare the current shortest_entry with what was identified in the db_tokens
+                                    // updating the variable in the outer scope with the smallest value
+                                    shortest_entry = Math.min(shortest_entry, db_tokens.min{ it.size() }.size())
+                                    new Tuple(db_tokens, key) }
+
+        def DB_TOKES_POS = 0
+        def DB_KEY = 1
+
+        // Remove characters that are shorter then the shortest database token as they cannot be part of a match
+        // This eliminates tokens that are part of a taxonomic string like `_s` or `spp`
+        species_data = species_data.findAll { it.size() >= shortest_entry }
+
+        def DB_SELECTION_SCORE_POS = 1
+        def DB_DATA_POS = 0
+        def MINIMUM_MATCH_SCORE = 1.0
+        def matched_databases = databases.collect {
+            db ->
+                def match_size = db[DB_TOKES_POS].size()
+                def tokens = window_string(species_data, match_size)
```

Comment:
I added a further comment to the code explaining why this is nessecary. But I will put it here as well for continuity sake!

The species names that are passed in may match multiple allele schemes but we need to select the most specific scheme. E.g. Someone may have a generic `Campylobacter` scheme and ` Campylobacter jejuni` scheme. 

The way the module is setup is to select for the optimal match of the databases provided, so that you can have a scheme for `Campylobacter coli` e.g into `Campylobacter` while `Campylobacter jejuni` is selected by for the more specific scheme.

The windowing function converts an array of strings into an array of n-grams, the number of n-grams matches the number of tokens (e.g. words) in the database scheme.

This lets us identify an exact match for the value passed in.

Reply:
Thanks so much for the explanation and including this in the comments as well. This is a great solution 😄